File: apps/wallet-mobile/etridefi-bloc/src/services/DEXService.ts
import axios from 'axios';
import BigNumber from 'bignumber.js';
import {Network, NETWORK_CONFIG, WRAPPED_ETR} from './ContractAddresses';

// Swap quote from DEX aggregator
export interface SwapQuote {
  id: string;
  chain: Network;
  fromToken: string;
  toToken: string;
  fromAmount: string;
  toAmount: string;
  rate: string;
  priceImpact: number;
  route: string;
  estimatedGas: number;
  timestamp: Date;
}

// Swap transaction to be signed
export interface SwapTransaction {
  id: string;
  quote: SwapQuote;
  fromAddress: string;
  to: string;
  data: string;
  value: string;
  gasLimit: number;
  gasPrice: string | null;
  nonce: number | null;
}

// DEX route info
export interface DEXRoute {
  dex: string;
  name: string;
  logoURL: string | null;
}

// DEX Error types
export enum DEXErrorType {
  INVALID_URL = 'INVALID_URL',
  NETWORK_ERROR = 'NETWORK_ERROR',
  PARSE_ERROR = 'PARSE_ERROR',
  NO_LIQUIDITY = 'NO_LIQUIDITY',
  RATE_LIMITED = 'RATE_LIMITED',
  API_ERROR = 'API_ERROR',
  INSUFFICIENT_BALANCE = 'INSUFFICIENT_BALANCE',
  SLIPPAGE_TOO_HIGH = 'SLIPPAGE_TOO_HIGH',
}

export class DEXError extends Error {
  type: DEXErrorType;
  code?: number;

  constructor(type: DEXErrorType, message: string, code?: number) {
    super(message);
    this.type = type;
    this.code = code;
    this.name = 'DEXError';
  }
}

// DEX Service singleton
class DEXServiceClass {
  private currentQuote: SwapQuote | null = null;
  private isLoadingQuote = false;
  private quoteError: string | null = null;
  private listeners: Set<() => void> = new Set();

  // API endpoints
  private readonly oneInchBaseURL = 'https://api.1inch.dev/swap/v6.0';
  private readonly jupiterBaseURL = 'https://quote-api.jup.ag/v6';

  // API keys
  private readonly zeroXAPIKey = 'bbd6ae09-7c0b-47b9-b6b5-e2682fd7d5c4';
  private readonly oneInchAPIKey = 'bbd6ae09-7c0b-47b9-b6b5-e2682fd7d5c4';

  // Subscribe to state changes
  subscribe(listener: () => void): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private notifyListeners(): void {
    this.listeners.forEach(listener => listener());
  }

  // Get current state
  getState() {
    return {
      currentQuote: this.currentQuote,
      isLoadingQuote: this.isLoadingQuote,
      quoteError: this.quoteError,
    };
  }

  // Get swap quote for token exchange
  async getQuote(
    fromToken: string,
    toToken: string,
    amount: string,
    chain: Network,
    slippage: number = 0.5,
  ): Promise<SwapQuote> {
    this.isLoadingQuote = true;
    this.quoteError = null;
    this.notifyListeners();

    try {
      let quote: SwapQuote;

      if (chain === Network.SOLANA) {
        quote = await this.getJupiterQuote(fromToken, toToken, amount, slippage);
      } else {
        quote = await this.get1inchQuote(fromToken, toToken, amount, chain, slippage);
      }

      this.currentQuote = quote;
      this.isLoadingQuote = false;
      this.notifyListeners();
      return quote;
    } catch (error: any) {
      this.quoteError = error.message || 'Failed to get quote';
      this.isLoadingQuote = false;
      this.notifyListeners();
      throw error;
    }
  }

  // Build swap transaction for execution
  async buildSwapTransaction(
    quote: SwapQuote,
    fromAddress: string,
    slippage: number = 0.5,
  ): Promise<SwapTransaction> {
    if (quote.chain === Network.SOLANA) {
      return this.buildJupiterSwapTx(quote, fromAddress, slippage);
    } else {
      return this.build1inchSwapTx(quote, fromAddress, slippage);
    }
  }

  // Get available DEX routes for a token pair
  async getAvailableRoutes(
    fromToken: string,
    toToken: string,
    chain: Network,
  ): Promise<DEXRoute[]> {
    switch (chain) {
      case Network.ETHEREUM:
        return [
          {dex: 'Uniswap V3', name: 'Uniswap', logoURL: null},
          {dex: 'SushiSwap', name: 'SushiSwap', logoURL: null},
          {dex: 'Curve', name: 'Curve', logoURL: null},
        ];
      case Network.BNB_CHAIN:
        return [
          {dex: 'PancakeSwap V3', name: 'PancakeSwap', logoURL: null},
          {dex: 'BiSwap', name: 'BiSwap', logoURL: null},
        ];
      case Network.POLYGON:
        return [
          {dex: 'QuickSwap', name: 'QuickSwap', logoURL: null},
          {dex: 'Uniswap V3', name: 'Uniswap', logoURL: null},
        ];
      case Network.ARBITRUM:
        return [
          {dex: 'Uniswap V3', name: 'Uniswap', logoURL: null},
          {dex: 'Camelot', name: 'Camelot', logoURL: null},
          {dex: 'SushiSwap', name: 'SushiSwap', logoURL: null},
        ];
      case Network.SOLANA:
        return [
          {dex: 'Raydium', name: 'Raydium', logoURL: null},
          {dex: 'Orca', name: 'Orca', logoURL: null},
          {dex: 'Jupiter', name: 'Jupiter Aggregator', logoURL: null},
        ];
      default:
        return [];
    }
  }

  // Clear current quote
  clearQuote(): void {
    this.currentQuote = null;
    this.quoteError = null;
    this.notifyListeners();
  }

  // 1inch Integration (EVM Chains)
  private async get1inchQuote(
    fromToken: string,
    toToken: string,
    amount: string,
    chain: Network,
    slippage: number,
  ): Promise<SwapQuote> {
    const chainId = NETWORK_CONFIG[chain].chainId;
    const amountWei = new BigNumber(amount)
      .multipliedBy(new BigNumber(10).pow(18))
      .toFixed(0);

    // Use native token address for non-contract tokens
    const nativeAddress = '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE';
    const fromTokenAddress = fromToken.startsWith('0x') ? fromToken : nativeAddress;
    const toTokenAddress = toToken.startsWith('0x') ? toToken : nativeAddress;

    // If no API key, return simulated quote
    if (!this.oneInchAPIKey) {
      return this.simulatedEVMQuote(fromToken, toToken, amount, chain);
    }

    try {
      const response = await axios.get(
        `${this.oneInchBaseURL}/${chainId}/quote`,
        {
          params: {
            src: fromTokenAddress,
            dst: toTokenAddress,
            amount: amountWei,
          },
          headers: {
            Accept: 'application/json',
            Authorization: `Bearer ${this.oneInchAPIKey}`,
          },
          timeout: 15000,
        },
      );

      if (response.status === 429) {
        throw new DEXError(DEXErrorType.RATE_LIMITED, 'Rate limited, try again later');
      }

      const data = response.data;
      const toAmountStr = data.toAmount;
      const outputAmount = new BigNumber(toAmountStr)
        .dividedBy(new BigNumber(10).pow(18))
        .toString();
      const rate = new BigNumber(outputAmount).dividedBy(amount).toString();

      return {
        id: this.generateId(),
        chain,
        fromToken,
        toToken,
        fromAmount: amount,
        toAmount: outputAmount,
        rate,
        priceImpact: data.estimatedGas ? 0.1 : 0.0,
        route: '1inch Aggregator',
        estimatedGas: data.estimatedGas || 150000,
        timestamp: new Date(),
      };
    } catch (error: any) {
      // Fallback to simulated quote on error
      if (error instanceof DEXError) throw error;
      return this.simulatedEVMQuote(fromToken, toToken, amount, chain);
    }
  }

  private async build1inchSwapTx(
    quote: SwapQuote,
    fromAddress: string,
    slippage: number,
  ): Promise<SwapTransaction> {
    // For demo, return a simulated transaction
    return {
      id: this.generateId(),
      quote,
      fromAddress,
      to: '0x1111111254EEB25477B68fb85Ed929f73A960582', // 1inch router
      data: '0x...', // Swap calldata
      value: quote.fromToken.startsWith('0x') ? '0' : quote.fromAmount,
      gasLimit: quote.estimatedGas,
      gasPrice: null,
      nonce: null,
    };
  }

  // Jupiter Integration (Solana)
  private async getJupiterQuote(
    fromToken: string,
    toToken: string,
    amount: string,
    slippage: number,
  ): Promise<SwapQuote> {
    const solMint = 'So11111111111111111111111111111111111111112';
    const wETRMint = WRAPPED_ETR[Network.SOLANA] || '';

    const inputMint = fromToken === 'SOL' ? solMint : fromToken;
    const outputMint = toToken === 'SOL' ? solMint : toToken;

    // Convert to lamports (9 decimals)
    const amountLamports = new BigNumber(amount)
      .multipliedBy(new BigNumber(10).pow(9))
      .toFixed(0);

    try {
      const response = await axios.get(`${this.jupiterBaseURL}/quote`, {
        params: {
          inputMint,
          outputMint,
          amount: amountLamports,
          slippageBps: Math.round(slippage * 100),
        },
        headers: {
          Accept: 'application/json',
        },
        timeout: 15000,
      });

      if (response.status !== 200) {
        return this.simulatedSolanaQuote(fromToken, toToken, amount);
      }

      const data = response.data;
      const outAmount = new BigNumber(data.outAmount)
        .dividedBy(new BigNumber(10).pow(9))
        .toString();
      const rate = new BigNumber(outAmount).dividedBy(amount).toString();
      const priceImpact = parseFloat(data.priceImpactPct || '0');

      return {
        id: this.generateId(),
        chain: Network.SOLANA,
        fromToken,
        toToken,
        fromAmount: amount,
        toAmount: outAmount,
        rate,
        priceImpact,
        route: data.routePlan?.[0]?.swapInfo || 'Jupiter',
        estimatedGas: 5000,
        timestamp: new Date(),
      };
    } catch (error) {
      return this.simulatedSolanaQuote(fromToken, toToken, amount);
    }
  }

  private async buildJupiterSwapTx(
    quote: SwapQuote,
    fromAddress: string,
    slippage: number,
  ): Promise<SwapTransaction> {
    return {
      id: this.generateId(),
      quote,
      fromAddress,
      to: 'JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4', // Jupiter program
      data: '', // Base64 encoded transaction
      value: '0',
      gasLimit: 200000,
      gasPrice: null,
      nonce: null,
    };
  }

  // Simulated quotes for demo/fallback
  private simulatedEVMQuote(
    fromToken: string,
    toToken: string,
    amount: string,
    chain: Network,
  ): SwapQuote {
    const wETRAddress = WRAPPED_ETR[chain]?.toLowerCase() || '';
    let rate: string;

    if (toToken.toLowerCase().includes(wETRAddress)) {
      rate = '6.67'; // Buying wETR
    } else {
      rate = '0.15'; // Selling wETR
    }

    const toAmount = new BigNumber(amount).multipliedBy(rate).toString();

    return {
      id: this.generateId(),
      chain,
      fromToken,
      toToken,
      fromAmount: amount,
      toAmount,
      rate,
      priceImpact: 0.15,
      route: 'Simulated DEX Route',
      estimatedGas: 150000,
      timestamp: new Date(),
    };
  }

  private simulatedSolanaQuote(
    fromToken: string,
    toToken: string,
    amount: string,
  ): SwapQuote {
    const rate = '0.0015'; // wETR/SOL rate
    let toAmount: string;

    if (toToken === WRAPPED_ETR[Network.SOLANA]) {
      toAmount = new BigNumber(amount).dividedBy(rate).toString();
    } else {
      toAmount = new BigNumber(amount).multipliedBy(rate).toString();
    }

    const calculatedRate = new BigNumber(toAmount).dividedBy(amount).toString();

    return {
      id: this.generateId(),
      chain: Network.SOLANA,
      fromToken,
      toToken,
      fromAmount: amount,
      toAmount,
      rate: calculatedRate,
      priceImpact: 0.1,
      route: 'Raydium (Simulated)',
      estimatedGas: 5000,
      timestamp: new Date(),
    };
  }

  private generateId(): string {
    return 'xxxx-xxxx-xxxx'.replace(/x/g, () =>
      Math.floor(Math.random() * 16).toString(16),
    );
  }

  // Calculate minimum received with slippage
  calculateMinimumReceived(quote: SwapQuote, slippage: number): string {
    return new BigNumber(quote.toAmount)
      .multipliedBy(1 - slippage / 100)
      .toString();
  }

  // Format swap preview
  formatSwapPreview(quote: SwapQuote): string {
    const from = new BigNumber(quote.fromAmount).toFormat(6);
    const to = new BigNumber(quote.toAmount).toFormat(6);
    return `${from} -> ${to}`;
  }

  // Check if price impact is high
  isHighPriceImpact(quote: SwapQuote): boolean {
    return quote.priceImpact > 1.0;
  }

  // Format rate for display
  formatRate(quote: SwapQuote): string {
    return new BigNumber(quote.rate).toFormat(6);
  }

  // Format price impact
  formatPriceImpact(quote: SwapQuote): string {
    return `${quote.priceImpact.toFixed(2)}%`;
  }
}

// Export singleton
export const DEXService = new DEXServiceClass();

File: apps/EtridefiBloc/Sources/EtridWalletSwift/Services/DEXService.swift
import Foundation
import Combine

// MARK: - DEX Aggregation Service

/// Service for fetching swap quotes and building transactions across DEXes
@MainActor
public class DEXService: ObservableObject {
    public static let shared = DEXService()

    // MARK: - Published Properties

    @Published public var currentQuote: SwapQuote?
    @Published public var isLoadingQuote = false
    @Published public var quoteError: String?

    // MARK: - Private Properties

    private let session: URLSession

    // API Endpoints
    private let oneInchBaseURL = "https://api.1inch.dev/swap/v6.0"
    private let jupiterBaseURL = "https://quote-api.jup.ag/v6"

    // 0x API key for Ëtridefi Bloc
    private let zeroXAPIKey = "bbd6ae09-7c0b-47b9-b6b5-e2682fd7d5c4"

    // 1inch API key (using 0x key for now - can be replaced with actual 1inch key)
    private let oneInchAPIKey = "bbd6ae09-7c0b-47b9-b6b5-e2682fd7d5c4"

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 15
        self.session = URLSession(configuration: config)
    }

    // MARK: - Public Methods

    /// Get swap quote for token exchange
    public func getQuote(
        fromToken: String,
        toToken: String,
        amount: Decimal,
        chain: ContractAddresses.Network,
        slippage: Double = 0.5
    ) async throws -> SwapQuote {
        isLoadingQuote = true
        quoteError = nil

        defer { isLoadingQuote = false }

        do {
            let quote: SwapQuote
            if chain == .solana {
                quote = try await getJupiterQuote(
                    fromToken: fromToken,
                    toToken: toToken,
                    amount: amount,
                    slippage: slippage
                )
            } else {
                quote = try await get1inchQuote(
                    fromToken: fromToken,
                    toToken: toToken,
                    amount: amount,
                    chain: chain,
                    slippage: slippage
                )
            }

            currentQuote = quote
            return quote
        } catch {
            quoteError = error.localizedDescription
            throw error
        }
    }

    /// Build swap transaction for execution
    public func buildSwapTransaction(
        quote: SwapQuote,
        fromAddress: String,
        slippage: Double = 0.5
    ) async throws -> SwapTransaction {
        if quote.chain == .solana {
            return try await buildJupiterSwapTx(quote: quote, fromAddress: fromAddress, slippage: slippage)
        } else {
            return try await build1inchSwapTx(quote: quote, fromAddress: fromAddress, slippage: slippage)
        }
    }

    /// Get available DEX routes for a token pair
    public func getAvailableRoutes(
        fromToken: String,
        toToken: String,
        chain: ContractAddresses.Network
    ) async throws -> [DEXRoute] {
        // Return known DEXes for each chain
        switch chain {
        case .ethereum:
            return [
                DEXRoute(dex: "Uniswap V3", name: "Uniswap", logoURL: nil),
                DEXRoute(dex: "SushiSwap", name: "SushiSwap", logoURL: nil),
                DEXRoute(dex: "Curve", name: "Curve", logoURL: nil)
            ]
        case .bnbChain:
            return [
                DEXRoute(dex: "PancakeSwap V3", name: "PancakeSwap", logoURL: nil),
                DEXRoute(dex: "BiSwap", name: "BiSwap", logoURL: nil)
            ]
        case .polygon:
            return [
                DEXRoute(dex: "QuickSwap", name: "QuickSwap", logoURL: nil),
                DEXRoute(dex: "Uniswap V3", name: "Uniswap", logoURL: nil)
            ]
        case .arbitrum:
            return [
                DEXRoute(dex: "Uniswap V3", name: "Uniswap", logoURL: nil),
                DEXRoute(dex: "Camelot", name: "Camelot", logoURL: nil),
                DEXRoute(dex: "SushiSwap", name: "SushiSwap", logoURL: nil)
            ]
        case .solana:
            return [
                DEXRoute(dex: "Raydium", name: "Raydium", logoURL: nil),
                DEXRoute(dex: "Orca", name: "Orca", logoURL: nil),
                DEXRoute(dex: "Jupiter", name: "Jupiter Aggregator", logoURL: nil)
            ]
        default:
            return []
        }
    }

    // MARK: - 1inch Integration (EVM Chains)

    private func get1inchQuote(
        fromToken: String,
        toToken: String,
        amount: Decimal,
        chain: ContractAddresses.Network,
        slippage: Double
    ) async throws -> SwapQuote {
        let chainId = chain.chainId
        let amountWei = (amount * pow(10, 18)).description.split(separator: ".").first.map(String.init) ?? "0"

        // Use 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE for native tokens
        let fromTokenAddress = fromToken.hasPrefix("0x") ? fromToken : "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"
        let toTokenAddress = toToken.hasPrefix("0x") ? toToken : "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"

        guard var components = URLComponents(string: "\(oneInchBaseURL)/\(chainId)/quote") else {
            throw DEXError.invalidURL
        }

        components.queryItems = [
            URLQueryItem(name: "src", value: fromTokenAddress),
            URLQueryItem(name: "dst", value: toTokenAddress),
            URLQueryItem(name: "amount", value: amountWei)
        ]

        guard let url = components.url else {
            throw DEXError.invalidURL
        }

        var request = URLRequest(url: url)
        request.addValue("application/json", forHTTPHeaderField: "Accept")

        if !oneInchAPIKey.isEmpty {
            request.addValue("Bearer \(oneInchAPIKey)", forHTTPHeaderField: "Authorization")
        }

        // For demo purposes, return a simulated quote if API key is not set
        if oneInchAPIKey.isEmpty {
            return simulatedEVMQuote(fromToken: fromToken, toToken: toToken, amount: amount, chain: chain)
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DEXError.networkError
        }

        if httpResponse.statusCode == 429 {
            throw DEXError.rateLimited
        }

        guard httpResponse.statusCode == 200 else {
            throw DEXError.apiError(httpResponse.statusCode)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DEXError.parseError
        }

        guard let toAmountStr = json["toAmount"] as? String,
              let toAmount = Decimal(string: toAmountStr) else {
            throw DEXError.parseError
        }

        let outputAmount = toAmount / pow(10, 18)
        let rate = outputAmount / amount
        let priceImpact = (json["estimatedGas"] as? Double).map { _ in 0.1 } ?? 0.0 // Simplified

        return SwapQuote(
            id: UUID().uuidString,
            chain: chain,
            fromToken: fromToken,
            toToken: toToken,
            fromAmount: amount,
            toAmount: outputAmount,
            rate: rate,
            priceImpact: priceImpact,
            route: "1inch Aggregator",
            estimatedGas: json["estimatedGas"] as? Int ?? 150000,
            timestamp: Date()
        )
    }

    private func build1inchSwapTx(
        quote: SwapQuote,
        fromAddress: String,
        slippage: Double
    ) async throws -> SwapTransaction {
        // For demo, return a simulated transaction
        return SwapTransaction(
            id: UUID().uuidString,
            quote: quote,
            fromAddress: fromAddress,
            to: "0x1111111254EEB25477B68fb85Ed929f73A960582", // 1inch router
            data: "0x...", // Swap calldata
            value: quote.fromToken.hasPrefix("0x") ? "0" : quote.fromAmount.description,
            gasLimit: quote.estimatedGas,
            gasPrice: nil, // Use network default
            nonce: nil // Fetch from network
        )
    }

    // MARK: - Jupiter Integration (Solana)

    private func getJupiterQuote(
        fromToken: String,
        toToken: String,
        amount: Decimal,
        slippage: Double
    ) async throws -> SwapQuote {
        // SOL mint address
        let solMint = "So11111111111111111111111111111111111111112"
        let wETRMint = ContractAddresses.wrappedETR[.solana] ?? ""

        let inputMint = fromToken == "SOL" ? solMint : fromToken
        let outputMint = toToken == "SOL" ? solMint : toToken

        // Convert to lamports (9 decimals for SOL, 9 for wETR on Solana)
        let scaledAmount = amount * pow(10, 9)
        let amountLamports = Int(truncating: scaledAmount as NSDecimalNumber)

        guard var components = URLComponents(string: "\(jupiterBaseURL)/quote") else {
            throw DEXError.invalidURL
        }

        components.queryItems = [
            URLQueryItem(name: "inputMint", value: inputMint),
            URLQueryItem(name: "outputMint", value: outputMint),
            URLQueryItem(name: "amount", value: String(amountLamports)),
            URLQueryItem(name: "slippageBps", value: String(Int(slippage * 100)))
        ]

        guard let url = components.url else {
            throw DEXError.invalidURL
        }

        var request = URLRequest(url: url)
        request.addValue("application/json", forHTTPHeaderField: "Accept")

        do {
            let (data, response) = try await session.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                // Fallback to simulated quote
                return simulatedSolanaQuote(fromToken: fromToken, toToken: toToken, amount: amount)
            }

            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let outAmountStr = json["outAmount"] as? String,
                  let outAmount = Decimal(string: outAmountStr) else {
                throw DEXError.parseError
            }

            let outputAmount = outAmount / pow(10, 9)
            let rate = outputAmount / amount
            let priceImpact = (json["priceImpactPct"] as? String).flatMap { Double($0) } ?? 0.0

            return SwapQuote(
                id: UUID().uuidString,
                chain: .solana,
                fromToken: fromToken,
                toToken: toToken,
                fromAmount: amount,
                toAmount: outputAmount,
                rate: rate,
                priceImpact: priceImpact,
                route: (json["routePlan"] as? [[String: Any]])?.first?["swapInfo"] as? String ?? "Jupiter",
                estimatedGas: 5000, // Solana compute units
                timestamp: Date()
            )
        } catch {
            // Fallback to simulated quote
            return simulatedSolanaQuote(fromToken: fromToken, toToken: toToken, amount: amount)
        }
    }

    private func buildJupiterSwapTx(
        quote: SwapQuote,
        fromAddress: String,
        slippage: Double
    ) async throws -> SwapTransaction {
        // In production, call Jupiter's /swap endpoint
        return SwapTransaction(
            id: UUID().uuidString,
            quote: quote,
            fromAddress: fromAddress,
            to: "JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4", // Jupiter program
            data: "", // Base64 encoded transaction
            value: "0",
            gasLimit: 200000,
            gasPrice: nil,
            nonce: nil
        )
    }

    // MARK: - Simulated Quotes (Demo)

    private func simulatedEVMQuote(
        fromToken: String,
        toToken: String,
        amount: Decimal,
        chain: ContractAddresses.Network
    ) -> SwapQuote {
        // Simulate exchange rates based on known prices
        let rate: Decimal
        if toToken.lowercased().contains(ContractAddresses.wrappedETR[chain]?.lowercased() ?? "") {
            // Buying wETR with native token
            rate = Decimal(string: "6.67")! // ~$0.15 wETR, assuming native ~$1
        } else {
            // Selling wETR
            rate = Decimal(string: "0.15")!
        }

        return SwapQuote(
            id: UUID().uuidString,
            chain: chain,
            fromToken: fromToken,
            toToken: toToken,
            fromAmount: amount,
            toAmount: amount * rate,
            rate: rate,
            priceImpact: 0.15,
            route: "Simulated DEX Route",
            estimatedGas: 150000,
            timestamp: Date()
        )
    }

    private func simulatedSolanaQuote(
        fromToken: String,
        toToken: String,
        amount: Decimal
    ) -> SwapQuote {
        let rate = Decimal(string: "0.0015")! // wETR/SOL rate (~$0.15 / ~$100)

        let outputAmount: Decimal
        if toToken == ContractAddresses.wrappedETR[.solana] {
            outputAmount = amount / rate // SOL -> wETR
        } else {
            outputAmount = amount * rate // wETR -> SOL
        }

        return SwapQuote(
            id: UUID().uuidString,
            chain: .solana,
            fromToken: fromToken,
            toToken: toToken,
            fromAmount: amount,
            toAmount: outputAmount,
            rate: outputAmount / amount,
            priceImpact: 0.1,
            route: "Raydium (Simulated)",
            estimatedGas: 5000,
            timestamp: Date()
        )
    }
}

// MARK: - Supporting Types

public struct SwapQuote: Identifiable, Codable {
    public let id: String
    public let chain: ContractAddresses.Network
    public let fromToken: String
    public let toToken: String
    public let fromAmount: Decimal
    public let toAmount: Decimal
    public let rate: Decimal
    public let priceImpact: Double
    public let route: String
    public let estimatedGas: Int
    public let timestamp: Date

    public var formattedRate: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.maximumFractionDigits = 6
        return formatter.string(from: rate as NSDecimalNumber) ?? "0"
    }

    public var formattedPriceImpact: String {
        return String(format: "%.2f%%", priceImpact)
    }

    public var isHighPriceImpact: Bool {
        priceImpact > 1.0
    }
}

public struct SwapTransaction: Identifiable, Codable {
    public let id: String
    public let quote: SwapQuote
    public let fromAddress: String
    public let to: String
    public let data: String
    public let value: String
    public let gasLimit: Int
    public let gasPrice: String?
    public let nonce: Int?
}

public struct DEXRoute: Identifiable, Codable {
    public var id: String { dex }
    public let dex: String
    public let name: String
    public let logoURL: String?
}

public enum DEXError: LocalizedError {
    case invalidURL
    case networkError
    case parseError
    case noLiquidity
    case rateLimited
    case apiError(Int)
    case insufficientBalance
    case slippageTooHigh

    public var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid API URL"
        case .networkError:
            return "Network error occurred"
        case .parseError:
            return "Failed to parse response"
        case .noLiquidity:
            return "Insufficient liquidity for this trade"
        case .rateLimited:
            return "Rate limited, try again later"
        case .apiError(let code):
            return "API error: \(code)"
        case .insufficientBalance:
            return "Insufficient balance"
        case .slippageTooHigh:
            return "Price impact too high"
        }
    }
}

// MARK: - Swap Preview Helper

public extension DEXService {

    /// Calculate minimum received amount with slippage
    func calculateMinimumReceived(quote: SwapQuote, slippage: Double) -> Decimal {
        return quote.toAmount * Decimal(1 - slippage / 100)
    }

    /// Format swap preview for display
    func formatSwapPreview(quote: SwapQuote) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .decimal
        formatter.maximumFractionDigits = 6

        let from = formatter.string(from: quote.fromAmount as NSDecimalNumber) ?? "0"
        let to = formatter.string(from: quote.toAmount as NSDecimalNumber) ?? "0"

        return "\(from) -> \(to)"
    }
}
